tensorflow/python/kernel_tests/sparse_tensor_dense_matmul_op_base.py

```diff
@@ -0,0 +1,254 @@
+# Copyright 2015 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Functional tests for SparseTensorDenseMatMul."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import time
+
+import numpy as np
+
+from tensorflow.python.client import session
+from tensorflow.python.framework import config
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import errors
+from tensorflow.python.framework import ops
+from tensorflow.python.framework import sparse_tensor
+from tensorflow.python.framework import test_util
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import sparse_ops
+from tensorflow.python.platform import app
+from tensorflow.python.platform import test
+
+
+def _maybe_complex(x):
+  if x.dtype.kind == "c":  # complex
+    return (x + 1j * x) / 2
+  return x
+
+
+class SparseTensorDenseMatMulTestBase(test.TestCase):
+
+  def _testMatmul(self,
+                  x,
+                  y,
+                  adjoint_a=False,
+                  adjoint_b=False,
+                  indices_dtype=np.int64):
+    x_mat = np.matrix(x)
+    if adjoint_a:
+      x_mat = x_mat.H
+    y_mat = np.matrix(y)
+    if adjoint_b:
+      y_mat = y_mat.H
+
+    np_ans = x_mat * y_mat
+
+    x_indices = np.vstack(np.where(x)).astype(indices_dtype).T
+    x_values = x[np.where(x)]
+    x_shape = x.shape
+
+    with self.cached_session(use_gpu=True):
+
+      sp_x_value = sparse_tensor.SparseTensorValue(
+          indices=x_indices, values=x_values, dense_shape=x_shape)
+
+      if (self.__class__ !=  SparseTensorDenseMatMulTestBase and
```

Comment:
Nice solution to the problem of the `SparseTensorDenseMatmulDeterministicTest` not being available in this scope.

This solution is a little brittle however because in that it relies on `SparseTensorDenseMatmulTest` not inheriting from `SparseTensorDenseMatMulTestBase` (as `SparseTensorDenseMatmulDeterministicTest` does) but instead `SparseTensorDenseMatmulTest` being a direct reference to `SparseTensorDenseMatMulTestBase`.

An alternative way of checking if this code is running inside an instance of the `SparseTensorDenseMatmulDeterministicTest` class, without having access to its token, is

`self.__class__.__name__ == "SparseTensorDenseMatmulDeterministicTest"`

Another option is to test for duck type:

`callable(getattr(self, "_getGpuDeterminismUnimplementedTypes", None))`